src/core/experimental/memory/memcpy.d

```diff
@@ -0,0 +1,401 @@
+ */
+static if (useSIMD)
+{
+
+import core.simd : float4;
+
+/**
+ * Handle Dynamic Types
+ * N.B.: While Dmemcpy's interface is C-like, it returns _nothing_.
+ *
+ * Params:
+ *  d = Pointer to memory destination to copy bytes to.
+ *  s = Pointer to memory source to copy bytes from.
+ *  n = Number of bytes to copy.
+ */
+void Dmemcpy(void* d, const(void)* s, size_t n)
+{
+    if (n <= 128)
+    {
+        Dmemcpy_small(d, s, n);
+    }
+    else
+    {
+        Dmemcpy_large(d, s, n);
+    }
+}
+
+/* Handle dynamic sizes <= 128. `d` and `s` must not overlap.
+ */
+private void Dmemcpy_small(void* d, const(void)* s, size_t n)
+{
+    if (n < 16) {
+        if (n & 0x01)
+        {
+            *(cast(ubyte*) d) = *(cast(const ubyte*) s);
+            ++d;
+            ++s;
+        }
+        if (n & 0x02)
+        {
+            *(cast(ushort*) d) = *(cast(const ushort*) s);
+            d += 2;
+            s += 2;
+        }
+        if (n & 0x04)
+        {
+            *(cast(uint*) d) = *(cast(const uint*) s);
+            d += 4;
+            s += 4;
+        }
+        if (n & 0x08)
+        {
+            *(cast(ulong*) d) = *(cast(const ulong*) s);
+        }
+        return;
+    }
+    if (n <= 32)
+    {
+        float4 xmm0 = load16fSSE(s);
+        float4 xmm1 = load16fSSE(s-16+n);
+        store16fSSE(d, xmm0);
+        store16fSSE(d-16+n, xmm1);
+        return;
+    }
+    // NOTE(stefanos): I'm writing using load/storeUnaligned() but you possibly can
+    // achieve greater performance using naked ASM. Be careful that you should either use
+    // only D or only naked ASM.
+    if (n <= 64)
+    {
+        float4 xmm0 = load16fSSE(s);
+        float4 xmm1 = load16fSSE(s+16);
+        float4 xmm2 = load16fSSE(s-32+n);
+        float4 xmm3 = load16fSSE(s-32+n+16);
+        store16fSSE(d, xmm0);
+        store16fSSE(d+16, xmm1);
+        store16fSSE(d-32+n, xmm2);
+        store16fSSE(d-32+n+16, xmm3);
+        return;
+    }
+    import core.simd : void16;
+    lstore64fSSE(d, s);
+    // Requires _no_ overlap.
+    n -= 64;
+    s = s + n;
+    d = d + n;
+    lstore64fSSE(d, s);
+}
+
+/* Handle dynamic sizes > 128. `d` and `s` must not overlap.
+ */
+private void Dmemcpy_large(void* d, const(void)* s, size_t n)
+{
+    // NOTE(stefanos): Alternative - Reach 64-byte
+    // (cache-line) alignment and use rep movsb
+    // Good for bigger sizes and only for Intel.
+
+    // Align destination (write) to 32-byte boundary
+    // NOTE(stefanos): We're using SSE, which needs 16-byte alignment.
+    // But actually, 32-byte alignment was quite faster (probably because
+    // the loads / stores are faster and there's the bottleneck).
+    uint rem = cast(ulong) d & 15;
+    if (rem)
+    {
+        store16fSSE(d, load16fSSE(s));
+        s += 16 - rem;
+        d += 16 - rem;
+        n -= 16 - rem;
+    }
+
+    static string loop(string prefetchChoice)()
+    {
+        return
+        "
```

Comment:
Use token strings

Reply:
I can't find an easy way to interpolate the `prefetchChoice` with token strings.

Reply:
I can't somehow answer on the comment above, so:

```
Function should still be private or have a clear confusion free name
```
If it is private, then the whole discussion about providing a C-like has no meaning.
So, if you don't want a C-like interface then I make it private otherwise, what would be a confusion-free name?

Reply:
> can't find an easy way to interpolate the prefetchChoice with token strings.

```
q{ ...} ~ ...
```